File: 130. Surrounded Regions.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        This is a DFS with recursion, but can also be done by DFS with stack or BFS with queue
        """
        if not board:
            return
        h, w = len(board), len(board[0])
        for i in range(h):
            self.helper(board, i, 0)
            self.helper(board, i, w-1)
        for i in range(w):
            self.helper(board, 0, i)
            self.helper(board, h-1, i)
        
        for i in range(h):
            for j in range(w):
                if board[i][j] == 'D':
                    board[i][j] = 'O'
                else:
                    board[i][j] = 'X'
                        
    def helper(self, board, i, j):
        if i < 0 or i >= len(board):
            return
        if j < 0 or j >= len(board[0]):
            return
        if board[i][j] == 'O':
            board[i][j] = 'D'
            self.helper(board, i+1, j)
            self.helper(board, i, j+1)
            self.helper(board, i-1, j)
            self.helper(board, i, j-1)

class UF:
    def __init__(self, N):
        self.ids = [i for i in range(N)]
        
    def root(self, A):
        while A != self.ids[A]:
            self.ids[A] = self.ids[self.ids[A]]
            A = self.ids[A]
        return A
    
    def find(self, A, B):
        return self.root(A) == self.root(B)
    
    def union(self, A, B):
        i = self.root(A)
        j = self.root(B)
        self.ids[i] = j

class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        A Union Find solution, keep 'O's on the boundry connected to a dummy node and change other 'O's into 'X'
        """
        if not board:
            return 
        m = len(board)
        n = len(board[0])
        uf = UF(m*n+1)
        for i in range(m):
            for j in range(n):
                if (i == 0 or i == m-1 or j == 0 or j == n-1) and board[i][j] == 'O':
                    uf.union(n*i+j, m*n)
                elif board[i][j] == 'O':
                    if board[i+1][j] == 'O':
                        uf.union(i*n+j, (i+1)*n+j)
                    if board[i-1][j] == 'O':
                        uf.union(i*n+j, (i-1)*n+j)
                    if board[i][j+1] == 'O':
                        uf.union(i*n+j, i*n+j+1)
                    if board[i][j-1] == 'O':
                        uf.union(i*n+j, i*n+j-1)
        for i in range(m):
            for j in range(n):
                if board[i][j] == 'O' and not uf.find(i*n+j, m*n):
                    board[i][j] = 'X'
        

File: 130. Surrounded Regions.java
class Solution {
    char[][] board;
    int wid;
    int hei;
    public void solve(char[][] board) {
        if (board==null || board.length==0) {
            return;
        }
        this.board = board;
        hei = board.length;
        wid = board[0].length;
        for (int i=0; i<wid; i++) {
            check(i, 0);
            check(i, hei-1);
        }
        for (int i=0; i<hei; i++) {
            check(0, i);
            check(wid-1, i);
        }
        
        for (int i=0; i<hei; i++){
            for (int j=0; j<wid; j++) {
                if (board[i][j] == '#') {
                    board[i][j] = 'O';
                } else {
                    board[i][j] = 'X';
                }
            }
        }
    }
    private void check(int x, int y) {
        if (x < 0 || x >= wid || y < 0 || y >= hei) {
            return;
        }
        if (board[y][x] == 'O') {
            board[y][x] = '#';
            check(x + 1, y);
            check(x - 1, y);
            check(x, y + 1);
            check(x, y - 1);
        }
    }
}
/**
 * Union find - not very suitable with this problem
 */
class Solution {
    public void solve(char[][] board) {
        if (board==null || board.length==0) {
            return;
        }
        int row = board.length, col = board[0].length;
        UF uf = new UF(row*col+1);
        int dummy = row*col;
        for (int i=0; i<row; i++) {
            for (int j=0; j<col; j++) {
                if (board[i][j] == 'O') {
                    if (i==0 || i==row-1 || j==0 || j==col-1) {
                        uf.union(i*col+j, dummy);
                    } else {
                        if (board[i-1][j] == 'O') {
                            uf.union(i*col+j, i*col-col+j);
                        }
                        if (board[i][j-1] == 'O') {
                            uf.union(i*col+j, i*col+j-1);
                        }
                        if (board[i+1][j] == 'O') {
                            uf.union(i*col+j, i*col+col+j);
                        }
                        if (board[i][j+1] == 'O') {
                            uf.union(i*col+j, i*col+j+1);
                        }
                    }
                }
            }
        }
        for (int i=0; i<row; i++) {
            for (int j=0; j<col; j++) {
                if (board[i][j] == 'O' && !uf.find(i*col+j, dummy)) {
                    board[i][j] = 'X';
                }
            }
        }
    }
}

class UF {
    int[] ids;
    public UF(int n) {
        ids = new int[n];
        for (int i=0; i<n; i++) {
            ids[i] = i;
        }
    }
    public int root(int id) {
        while (ids[id] != id) {
            ids[id] = ids[ids[id]];
            id = ids[id];
        }
        return id;
    }
    public boolean find(int A, int B) {
        return root(A) == root(B);
    }
    public void union(int A, int B) {
        int x = root(A);
        int y = root(B);
        ids[x] = y;
    }
}